File: Project Academy/9. Room/java/app/src/main/java/com/dicoding/academies/ui/reader/CourseReaderViewModel.java
package com.dicoding.academies.ui.reader;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;
import androidx.lifecycle.ViewModel;

import com.dicoding.academies.data.AcademyRepository;
import com.dicoding.academies.data.source.local.entity.ModuleEntity;
import com.dicoding.academies.vo.Resource;

import java.util.List;


public class CourseReaderViewModel extends ViewModel {

    private MutableLiveData<String> courseId = new MutableLiveData<>();
    private MutableLiveData<String> moduleId = new MutableLiveData<>();
    private AcademyRepository academyRepository;

    public CourseReaderViewModel(AcademyRepository mAcademyRepository) {
        this.academyRepository = mAcademyRepository;
    }

    public LiveData<Resource<List<ModuleEntity>>> modules = Transformations.switchMap(courseId,
        mCourseId -> academyRepository.getAllModulesByCourse(mCourseId));


    public void setCourseId(String courseId) {
        this.courseId.setValue(courseId);
    }

    public LiveData<Resource<ModuleEntity>> selectedModule = Transformations.switchMap(moduleId,
            selectedPosition -> academyRepository.getContent(selectedPosition)
    );

    public void setSelectedModule(String moduleId) {
        this.moduleId.setValue(moduleId);
    }

    public void readContent(ModuleEntity module) {
        academyRepository.setReadModule(module);
    }

    public int getModuleSize() {
        if (modules.getValue() != null) {
            List<ModuleEntity> moduleEntities = modules.getValue().data;
            if (moduleEntities != null) {
                return moduleEntities.size();
            }
        }
        return 0;
    }

    public void setNextPage() {
        if (selectedModule.getValue() != null && modules.getValue() != null) {
            ModuleEntity moduleEntity = selectedModule.getValue().data;
            List<ModuleEntity> moduleEntities = modules.getValue().data;
            if (moduleEntity != null && moduleEntities != null) {
                int position = moduleEntity.getPosition();
                if (position < moduleEntities.size() && position >= 0) {
                    setSelectedModule(moduleEntities.get(position + 1).getModuleId());
                }
            }
        }
    }

    public void setPrevPage() {
        if (selectedModule.getValue() != null && modules.getValue() != null) {
            ModuleEntity moduleEntity = selectedModule.getValue().data;
            List<ModuleEntity> moduleEntities = modules.getValue().data;
            if (moduleEntity != null && moduleEntities != null) {
                int position = moduleEntity.getPosition();
                if (position < moduleEntities.size() && position >= 0) {
                    setSelectedModule(moduleEntities.get(position - 1).getModuleId());
                }
            }
        }
    }
}

File: Project Academy/12. Pagination-test/kotlin/app/src/main/java/com/dicoding/academies/ui/reader/CourseReaderViewModel.kt
package com.dicoding.academies.ui.reader

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Transformations
import androidx.lifecycle.ViewModel
import com.dicoding.academies.data.AcademyRepository
import com.dicoding.academies.data.source.local.entity.ModuleEntity
import com.dicoding.academies.vo.Resource


class CourseReaderViewModel(private val academyRepository: AcademyRepository) : ViewModel() {

    var courseId = MutableLiveData<String>()

    var moduleId = MutableLiveData<String>()

    fun setSelectedCourse(courseId: String) {
        this.courseId.value = courseId
    }

    fun setSelectedModule(moduleId: String) {
        this.moduleId.value = moduleId
    }

    var modules: LiveData<Resource<List<ModuleEntity>>> = Transformations.switchMap(courseId) { mCourseId ->
        academyRepository.getAllModulesByCourse(mCourseId)
    }

    var selectedModule: LiveData<Resource<ModuleEntity>> = Transformations.switchMap(moduleId) { selectedPosition ->
        academyRepository.getContent(selectedPosition)
    }

    fun readContent(module: ModuleEntity) {
        academyRepository.setReadModule(module)
    }

    fun getModuleSize(): Int {
        if (modules.value != null) {
            val moduleEntities = modules.value?.data
            if (moduleEntities != null) {
                return moduleEntities.size
            }
        }
        return 0
    }

    fun setNextPage() {
        if (selectedModule.value != null && modules.value != null) {
            val moduleEntity = selectedModule.value?.data
            val moduleEntities = modules.value?.data
            if (moduleEntity != null && moduleEntities != null) {
                val position = moduleEntity.position
                if (position < moduleEntities.size && position >= 0) {
                    moduleId.value = moduleEntities[position + 1].moduleId
                }
            }
        }
    }

    fun setPrevPage() {
        if (selectedModule.value != null && modules.value != null) {
            val moduleEntity = selectedModule.value?.data
            val moduleEntities = modules.value?.data
            if (moduleEntity != null && moduleEntities != null) {
                val position = moduleEntity.position
                if (position < moduleEntities.size && position >= 0) {
                    moduleId.value = moduleEntities[position - 1].moduleId
                }
            }
        }
    }
}

